src/Shared/TagTransformerJsonHelper.cs

```diff
@@ -0,0 +1,190 @@
+            string[] stringArray => WriteToString(stringArray, WriteStringValue),
+            bool[] booleanArray => WriteToString(booleanArray, WriteBooleanValue),
+
+            // Runtime allows casting byte[] to sbyte[] and vice versa, so pattern match to sbyte[] doesn't work since it goes through byte[]
+            // Similarly (array is sbyte[]) doesn't help either. Only real type comparison works.
+            // This is true for byte/sbyte, short/ushort, int/uint and so on
+            byte[] byteArray => array.GetType() == typeof(sbyte[]) ? WriteToString((sbyte[])array, WriteSByteValue) : WriteToString(byteArray, WriteByteValue),
+            short[] shortArray => array.GetType() == typeof(ushort[]) ? WriteToString((ushort[])array, WriteUShortValue) : WriteToString(shortArray, WriteShortValue),
+            int[] intArray => array.GetType() == typeof(uint[]) ? WriteToString((uint[])array, WriteUIntValue) : WriteToString(intArray, WriteIntValue),
+
+            long[] longArray => WriteToString(longArray, WriteLongValue),
+            float[] floatArray => WriteToString(floatArray, WriteFloatValue),
+            double[] doubleArray => WriteToString(doubleArray, WriteDoubleValue),
+
+            _ => throw new NotSupportedException($"Unexpected serialization of array of type {array.GetType()}."),
+        };
+
+    private static void WriteCharValue(char v, Utf8JsonWriter writer)
+    {
+        Span<char> charSpan = stackalloc char[1];
+        charSpan[0] = v;
+        writer.WriteStringValue(charSpan);
+    }
+
+    private static void WriteStringValue(string v, Utf8JsonWriter writer) => writer.WriteStringValue(v);
+
+    private static void WriteBooleanValue(bool v, Utf8JsonWriter writer) => writer.WriteBooleanValue(v);
+
+    private static void WriteByteValue(byte v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static void WriteSByteValue(sbyte v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static void WriteShortValue(short v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static void WriteUShortValue(ushort v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static void WriteIntValue(int v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static void WriteUIntValue(uint v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static void WriteLongValue(long v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static void WriteFloatValue(float v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static void WriteDoubleValue(double v, Utf8JsonWriter writer) => writer.WriteNumberValue(v);
+
+    private static string WriteToString<T>(T[] data, Action<T, Utf8JsonWriter> writeAction)
+    {
+        // Some rough estimate of the necessary size. We estimate each element to be 10 characters long (plus the separator)
```

Comment:
(minor question): would it make sense for the type-specific code to pass this "hint" in to the method? For example `ushort` can only be 5 chars, right? 0-65535.

Reply:
Good suggestion - added